src/types/entities/schelling.ts

```diff
@@ -37,72 +37,142 @@ export class SchellingGame {
 	// All players have a unique overlay address
 	private players: BTree_<string, Player>
 	private rounds: BTree_<number, Round>
+	private currentRoundNo = 0
+	private lastBlock: BlockDetails
 
 	private myOverlays: string[] = []
 	private myAccounts: string[] = []
 
+	private static _showLogs = false
+
 	/**
 	 * Make the constructor private so that it can't be called.
 	 * @private
 	 */
 	private constructor() {
 		this.players = new BTree()
 		this.rounds = new BTree()
+		this.currentRoundNo = 0 // We haven't started the game yet
+		this.lastBlock = { blockNo: 0, blockTimestamp: 0 }
+	}
+
+	public highlightOverlay(overlay: string) {
+		if (!this.myOverlays.includes(overlay)) {
+			this.myOverlays[this.myOverlays.length] = overlay
+			this.players.set(
+				overlay,
+				new Player(overlay, undefined, undefined, this.size)
+			)
+		}
 	}
 
 	// --- game logic ---
 
-	// --- commit
-	public commit(overlay: string, block: BlockDetails) {
+	public newBlock(block: BlockDetails): string {
 		const roundNo = SchellingGame.roundFromBlockNo(block.blockNo)
+		if (roundNo != this.currentRoundNo) {
+			if (this.currentRoundNo != 0) {
+				const round = this.rounds.get(this.currentRoundNo)
+				if (round && !round.claim) {
+					round.lastBlock = this.lastBlock
+					round.unclaimed = true
+					round.render()
+				}
+			}
+			this.currentRoundNo = roundNo
+		}
+		this.lastBlock = block
+		const blocksPerRound = config.blocksPerRound // TODO use configured blocksPerRound
+		const offset = block.blockNo % blocksPerRound
+		let phase
+		let length
+		let elapsed
+		if (offset < blocksPerRound / 4) {
+			phase = 'commit'
+			length = blocksPerRound / 4
+			elapsed = offset + 1
+		} else if (offset <= blocksPerRound / 2) {
+			phase = 'reveal'
+			length = blocksPerRound / 4 + 1
+			elapsed = offset - blocksPerRound / 4 + 1
+		} else {
+			phase = 'claim'
+			length = blocksPerRound / 2 - 1
+			elapsed = offset - blocksPerRound / 2
+		}
+		const remaining = length - elapsed
+		const percent = Math.floor((elapsed * 100) / length)
+
+		let line = `${Round.roundString(
+			block.blockNo
+		)} ${percent}% of ${phase}, ${remaining} blocks left`
+		if (config.blocksPerRound - offset - 1 != remaining)
+			line = line + `, ${config.blocksPerRound - offset - 1} in round`
+		return line
+	}
 
+	public getOrCreateRound(roundNo: number, block: BlockDetails): Round {
 		// create the round if it doesn't exist
 		if (!this.rounds.has(roundNo)) {
 			this.rounds.set(roundNo, new Round(block))
-		} else {
-			// update the round timestamp
-			this.rounds.get(roundNo)!.lastBlock = block
+			this.players.forEachPair((overlay, player) => {
+				player.notPlaying()
+			})
```

Comment:
For determining if a player is playing or not, we have the invariant that their overlay is located in the `players` array in the `Round` entity. So, we can use the `round.players.includes(overlay)` as a determinant for the player playing or not relative to the specific round.

Reply:
Well, sort of.   The All Players box wants to highlight the current round players, but there is no actual definition of "current round" available to the Player.render().  That's why I did the notPlaying() the way I did to reduce the coupling back to the game from Player.  And it turns out the Players are nearly always in a Round somewhere, but the is/notPlaying flag really wants to track the current round, not just any Round.